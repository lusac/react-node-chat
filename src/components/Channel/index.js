import React from 'react'
import PropTypes from 'prop-types'

export default class Channel extends React.Component {
  constructor(props) {
    super()
    this.state = {
      msg: '',
      channel: props.channel || {},
      msgs: props.channel.msgs || []
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.channel.id !== prevProps.channel.id) {
      this.updateChannelConnection()
    }
  }

  updateChannelConnection() {
    this.setState({
      msgs: this.props.channel.msgs
    }, () => {
      this.props.socket.off('message').on('message', (msg) => {
        this.setState({
          msgs: [...this.state.msgs, msg]
        })
      })
    })
  }

  onMsgChange(e) {
    this.setState({
      msg: e.target.value
    })
  }

  onMsgSubmit(e) {
    e.preventDefault()
    this.props.socket.emit('message', {
      channelID: this.props.channel.id,
      msg: this.state.msg
    })
    this.setState({
      msg: ''
    })
  }

  render() {
    return (
      <div className="channel">
        <div className="channel__header">
          header
        </div>

        <div className="channel__section">
          <div className="channel__chat">
            {!this.props.channel.id && <span>Nenhum canal selecionado</span>}
            {this.props.channel.id && this.state.msgs.map((msg, i) => <div key={i}>{msg}</div> )}
          </div>

          {this.props.channel.id &&
            <div className="channel__chat-input__wrapper">
              <form onSubmit={this.onMsgSubmit.bind(this)}>
                <input
                  placeholder="escreva uma mensagem"
                  className="channel__chat-input"
                  value={this.state.msg}
                  onChange={this.onMsgChange.bind(this)} />
                {/* <input type="submit" value="Submit" /> */}
              </form>
            </div>
          }
        </div>
      </div>
    )
  }
}

Channel.propTypes = {
  channel: PropTypes.object,
  socket: PropTypes.object
}