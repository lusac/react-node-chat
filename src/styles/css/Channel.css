.room {
  background: #FFF;
}
.room__header {
  display: flex;
  align-items: center;
  justify-content: center;
  border-right: 1px solid #e1e1e1;
}
.room__section {
  overflow: auto;
  display: flex;
  flex-direction: column;
}
.room__messages {
  height: 75%;
  padding: 10px;
}
.room__chat-input {
  width: 85%;
  height: 100%;
  outline: none;
  padding: 10px;
}
.room__chat-input__wrapper {
  height: 25%;
  padding: 10px;
}

/*# sourceMappingURL=Channel.css.map */
