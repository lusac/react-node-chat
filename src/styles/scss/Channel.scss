.channel {

  &__empty-state {
    height: 100vh;
    display: flex;
    align-items: center;
    justify-content: center;

    strong {
      width: 50%;
      color: #333;
      font-size: 24px;
      max-width: 300px;
      font-family: arial;
    }
  }

  &__content {
    display: grid;
    height: 100vh;
    background: #FFF;
    grid-template-rows: 70px auto;
  }

  &__header {
    display: flex;
    padding: 0 20px;
    font-weight: bold;
    align-items: center;
    color: #00c876;
    box-shadow: 0px 2px 1px 0px #eaeaea;
  }

  &__section {
    padding: 20px;
    display: flex;
    overflow: auto;
    flex-direction: column;
  }

  &__messages {
    height: 75%;
    padding: 10px;
  }

  &__chat-input {
    width: 85%;
    height: 100%;
    outline: none;
    padding: 10px;

    &__wrapper {
      height: 25%;
      padding: 10px;
    }
  }
}