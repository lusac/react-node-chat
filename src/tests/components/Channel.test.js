/* global it describe expect beforeEach jest */
import React from 'react'
import { shallow, configure } from 'enzyme'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import Channel from '../../components/Channel'

configure({ adapter: new Adapter() })

describe('Channel component', () => {
  describe('snapshot', () => {
    it('channel not selected', () => {
      const
        channel = {},
        username = 'xunda',
        renderedValue =  renderer.create(<Channel channel={channel} username={username} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })

    it('channel already joined', () => {
      const
        channel = { id: 1, msgs: [], joined: true },
        username = 'xunda',
        renderedValue =  renderer.create(<Channel channel={channel} username={username} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })

    it('channel not joined', () => {
      const
        channel = { id: 1, msgs: [], joined: false },
        username = 'xunda',
        renderedValue =  renderer.create(<Channel channel={channel} username={username} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })

    it('channel selected with no msgs', () => {
      const
        channel = { id: 1, msgs: [] },
        username = 'xunda',
        renderedValue =  renderer.create(<Channel channel={channel} username={username} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })

    it('channel selected with msgs', () => {
      const
        channel = { id: 1, msgs: [{username: 'xunda', text: 'hello!', date: new Date('Mon Aug 27 2018 11:22:33')}] },
        username = 'xunda',
        renderedValue =  renderer.create(<Channel channel={channel} username={username} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })
  })

  describe('rendered', () => {
    describe('channel not selected', () => {
      let wrapper, channel = {}

      beforeEach(() => {
        wrapper = shallow(<Channel channel={channel}/>)
      })

      it('correct label', () => {
        expect(wrapper.find('.channel__empty-state strong').get(0).props.children).toBe('Escolha ou crie um canal para conversar com outras pessoas!')
      })
    })

    describe('channel selected with no msgs', () => {
      let wrapper, channel = { id: 1, name: 'channel 1', msgs: [] }

      beforeEach(() => {
        wrapper = shallow(<Channel channel={channel}/>)
      })

      it('correct label', () => {
        expect(wrapper.find('.channel__section>span').get(0).props.children).toBe('nenhuma mensagem')
      })
    })

    describe('channel selected with msgs', () => {
      let wrapper, channel = { id: 1, name: 'channel 1', msgs: [{text:'hello!', username: 'xunda', date: new Date('Mon Aug 27 2018 11:22:33')}] }

      beforeEach(() => {
        wrapper = shallow(<Channel channel={channel}/>)
      })

      it('correct label', () => {
        expect(wrapper.find('Message').length).toBe(1)
      })
    })

    describe('channel already joined', () => {
      let wrapper, channel = { id: 1, joined: true, name: 'channel 1', msgs: [{text:'hello!', username: 'xunda', date: new Date('Mon Aug 27 2018 11:22:33')}] }

      beforeEach(() => {
        wrapper = shallow(<Channel channel={channel}/>)
      })

      it('has input and dont have button to join', () => {
        expect(wrapper.find('.channel__chat-input').length).toBe(1)
        expect(wrapper.find('.channel__footer button').length).toBe(0)
      })
    })

    describe('channel not joined', () => {
      let wrapper, channel = { id: 1, joined: false, name: 'channel 1', msgs: [{text:'hello!', username: 'xunda', date: new Date('Mon Aug 27 2018 11:22:33')}] }

      beforeEach(() => {
        wrapper = shallow(<Channel channel={channel}/>)
      })

      it('has input and dont have button to join', () => {
        expect(wrapper.find('.channel__chat-input').length).toBe(0)
        expect(wrapper.find('.channel__footer button').length).toBe(1)
      })
    })
  })

  it('onMsgChange method', () => {
    let
      channel = { id: 1, name: 'channel 1', msgs: [] },
      wrapper = shallow(<Channel channel={channel}/>)

    expect(wrapper.instance().state.msg).toBe('')
    wrapper.instance().onMsgChange({target: {value: 'hello!'}})
    expect(wrapper.instance().state.msg).toBe('hello!')
  })

  it('onMsgSubmit method', () => {
    const d = new Date('2017-06-13T04:41:20')
    Date = class extends Date {
      constructor() {
        return d
      }
    }

    let
      e = { preventDefault: () => 'oie' },
      socket = { emit: (a, b) => (a, b) },
      user = {name: 'xunda', color: 2},
      channel = { id: 1, name: 'channel 1', msgs: [] },
      wrapper = shallow(<Channel user={user} socket={socket} channel={channel}/>)

    wrapper.instance().onMsgChange({target: {value: 'hello!'}})

    const spyPreventDefault = jest.spyOn(e, 'preventDefault')
    const spyEmit = jest.spyOn(wrapper.instance().props.socket, 'emit')

    wrapper.instance().onMsgSubmit(e)

    expect(spyPreventDefault).toHaveBeenCalled()
    expect(spyEmit).toHaveBeenCalledWith('message', {"channelID": 1, "msg": {"color": 2, "date": new Date(), "text": "hello!", "username": 'xunda'}})
    expect(wrapper.instance().state.msg).toBe('')
  })

  it('joinChannel method', () => {
    let
      socket = { emit: (a, b) => (a, b) },
      user = {name: 'xunda', color: 2},
      channel = { id: 1, name: 'channel 1', msgs: [] },
      wrapper = shallow(<Channel user={user} socket={socket} channel={channel}/>)

    const spy = jest.spyOn(wrapper.instance().props.socket, 'emit')
    wrapper.instance().joinChannel()
    expect(spy).toHaveBeenCalledWith('join channel', 1)
  })
})