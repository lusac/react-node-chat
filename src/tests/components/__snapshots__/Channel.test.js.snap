// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel component snapshot channel already joined 1`] = `
<div
  className="channel"
>
  <div
    className="channel__content"
  >
    <div
      className="channel__header"
    />
    <div
      className="channel__section"
    >
      <span>
        nenhuma mensagem
      </span>
    </div>
    <div
      className="channel__footer "
    >
      <form
        onSubmit={[Function]}
      >
        <input
          className="channel__chat-input"
          onChange={[Function]}
          placeholder="escreva uma mensagem"
          value=""
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`Channel component snapshot channel not joined 1`] = `
<div
  className="channel"
>
  <div
    className="channel__content"
  >
    <div
      className="channel__header"
    />
    <div
      className="channel__section"
    >
      <span>
        nenhuma mensagem
      </span>
    </div>
    <div
      className="channel__footer channel__footer--join"
    >
      <button
        className="blue"
        onClick={[Function]}
      >
        Entrar no canal
      </button>
    </div>
  </div>
</div>
`;

exports[`Channel component snapshot channel not selected 1`] = `
<div
  className="channel"
>
  <div
    className="channel__empty-state"
  >
    <strong>
      Escolha ou crie um canal para conversar com outras pessoas!
    </strong>
  </div>
</div>
`;

exports[`Channel component snapshot channel selected with msgs 1`] = `
<div
  className="channel"
>
  <div
    className="channel__content"
  >
    <div
      className="channel__header"
    />
    <div
      className="channel__section"
    >
      <div
        className="message "
      >
        <div
          className="message__left"
        >
          <span
            className="message__photo colorundefined"
          >
            X
          </span>
        </div>
        <div
          className="message__right"
        >
          <strong
            className="message__info"
          >
            <span>
              xunda
            </span>
            <span
              className="message__date"
            >
              8/27/2018, 11:22 AM
            </span>
          </strong>
          <span
            className="message__text"
          >
            hello!
          </span>
        </div>
      </div>
    </div>
    <div
      className="channel__footer channel__footer--join"
    >
      <button
        className="blue"
        onClick={[Function]}
      >
        Entrar no canal
      </button>
    </div>
  </div>
</div>
`;

exports[`Channel component snapshot channel selected with no msgs 1`] = `
<div
  className="channel"
>
  <div
    className="channel__content"
  >
    <div
      className="channel__header"
    />
    <div
      className="channel__section"
    >
      <span>
        nenhuma mensagem
      </span>
    </div>
    <div
      className="channel__footer channel__footer--join"
    >
      <button
        className="blue"
        onClick={[Function]}
      >
        Entrar no canal
      </button>
    </div>
  </div>
</div>
`;
